core/src/main/java/com/linecorp/armeria/common/stream/StreamMessageInputStream.java

```diff
@@ -0,0 +1,182 @@
+        if (inputStream == null || inputStream.available() == 0) {
+            try {
+                inputStream = subscriber.nextStream();
+            } catch (InterruptedException e) {
+                return -1;
+            }
+        }
+        return function.apply(inputStream);
+    }
+
+    @Override
+    public void close() throws IOException {
+        checkClosed();
+        closed = true;
+        subscriber.cancel();
+    }
+
+    @Override
+    public int available() throws IOException {
+        checkClosed();
+        if (inputStream == null) {
+            return 0;
+        }
+        return inputStream.available();
+    }
+
+    private void checkClosed() throws IOException {
+        if (closed) {
+            throw new IOException("Stream closed");
+        }
+    }
+
+    private static final class StreamMessageInputStreamSubscriber<T> implements Subscriber<T> {
+
+        private static final InputStream EMPTY_STREAM = new InputStream() {
+            @Override
+            public int read() {
+                return -1;
+            }
+        };
+
+        private final Function<? super T, ? extends HttpData> httpDataConverter;
+        private final CompletableFuture<Subscription> upstream = new CompletableFuture<>();
+        private final BlockingQueue<InputStream> queue = new LinkedBlockingDeque<>();
+        private volatile boolean closed;
+
+        StreamMessageInputStreamSubscriber(Function<? super T, ? extends HttpData> httpDataConverter) {
+            requireNonNull(httpDataConverter, "httpDataConverter");
+            this.httpDataConverter = httpDataConverter;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            upstream.complete(subscription);
+        }
+
+        @Override
+        public void onNext(T item) {
+            requireNonNull(item, "item");
+            if (closed) {
+                 StreamMessageUtil.closeOrAbort(item);
+                 return;
+            }
+            try {
+                queue.add(httpDataConverter.apply(item).toInputStream());
+            } catch (Throwable ex) {
+                StreamMessageUtil.closeOrAbort(item, ex);
+                upstream.join().cancel();
+                onError(ex);
+            }
+        }
+
+        @Override
+        public void onError(Throwable cause) {
+            closed = true;
+            queue.add(EMPTY_STREAM); // to wake the BlockingQueue
+        }
+
+        @Override
+        public void onComplete() {
+            closed = true;
+            queue.add(EMPTY_STREAM); // to wake the BlockingQueue
+        }
+
+        public InputStream nextStream() throws InterruptedException {
+            if (closed) {
+                return EMPTY_STREAM;
+            }
+            if (queue.isEmpty()) {
+                upstream.join().request(1);
+            }
+            return queue.take();
+        }
```

Comment:
![image](https://user-images.githubusercontent.com/34854527/156201218-ee45678a-71a0-4ee6-9ba7-6eef3f8b5bc9.png)

To wait `StreamMessage`'s next Item from `StreamMessageInputStream#read`, I used `BlockingQueue`!

Cause I'm not familiar with this codes (reactive streams..), feel free to give me any suggestion! Maybe I'm in wrong direction 😅
I refer to `FuseableStreamMessage` code a lot.